qualang_tools/addons/calibration/calibrations_API.py

```diff
@@ -0,0 +1,1114 @@
+
+        :param configuration:
+        :param readout: readout element and operation as defined in the configuration. Must be written as a tuple (element, operation).
+        :param qubit: qubit element and operation as defined in the configuration. Must be written as a tuple (element, operation).
+        :param integration_weights: integration weights as defined in the configuration. Must be written as a tuple of 4 items ('cos', 'sin', 'minus_sin', 'cos').
+        :param outputs: readout element outputs as defined in the configuration. Must be a python tuple of two items ('out1', 'out2').
+        :param flux_line: (optional) flux line element and operation as defined in the configuration. Must be written as a tuple (element, operation).
+        """
+        self.config = configuration
+        # Check the validity of the user inputs for elements, operations, int weights and outputs
+        if self._check_element(readout[0]):
+            self.readout_elmt = readout[0]
+        if self._check_element(qubit[0]):
+            self.qubit_elmt = qubit[0]
+        if flux_line is not None:
+            if self._check_element(flux_line[0]):
+                self.flux_line_elmt = flux_line[0]
+            if self._check_operation(flux_line[1], flux_line[0]):
+                self.flux_line_op = flux_line[1]
+        else:
+            self.flux_line_elmt = None
+        if self._check_operation(readout[1], readout[0]):
+            self.readout_op = readout[1]
+        if self._check_operation(qubit[1], qubit[0]):
+            self.qubit_op = qubit[1]
+        if self._check_iw(integration_weights):
+            self.iw = integration_weights
+        if self._check_outputs(outputs):
+            self.out = outputs
+
+        self.calibration_list = []
+        self.scan_var = {}
+        self.results = {}
+        self.cooldown_time = None
+        self.user_specifics = {}
+        self.simulation = False
+        self.qm = None
+        self.plot_options = {
+            "fontsize": 14,
+            "color": "b",
+            "marker": None,
+            "linewidth": 2,
+            "figsize": (12, 15),
+        }
+
+    def set_resonator_spectroscopy(
+        self, scan_variables, iterations=1, cooldown_time=None
+    ):
+        """Sets the resonator spectroscopy calibration.
+        This program will measure the resonator response for different readout pulse frequencies defined in scan_variables.
+
+        :param scan_variables: variables to be scanned. Must be an array of tuples [('variable', [values])]. 1D and 2D scans are available.
+        :param iterations: number of averaging iterations. Must be a positive Python integer.
+        :param cooldown_time: resonator or qubit cooldown time in clock cycles (4ns). Must be an integer.
+        :return: None
+        """
+        calibration = "Resonator_spec"
+        self._user_checks(calibration, scan_variables, iterations, cooldown_time)
```

Comment:
The function `_user_checks` sounds like it is only checking things, but in fact, it also adds the calibration to the list.
I would separate the two:
```
self._user_checks()
self._list.append()
```

Reply:
Also, you currently start with adding the calibration to the list, which means it'll be added even if the parameters are invalid.
It's better to first do the validation and then add it.
You can also just rename the function to `_add_calibration_to_list()` and just change the order inside.

Reply:
I don't know because this way I can directly append the variables to scan_var and results if it is valid. Otherwise I would have to duplicate the same loops into two separated functions... Do you think it is necessary or can I just rename the function to _user_checks_and_initialization()?

Reply:
Not sure I follow, let's discuss

Reply:
Now I loop over the variables and if it is valid I update the corresponding dictionaries. If I split into two functions, I have to loop over them first for the check and a second time to update the dictionaries in the initialization function, but you're right it may be cleaner, I was just lazy...